### PR TITLE
add teacher table

### PR DESCRIPTION
**Pull request by brian386**
## Notion ticket link

[Implement teacher users table](https://www.notion.so/uwblueprintexecs/Implement-Teacher-Users-table-eff0155d8e484d0dbcc5165a792482d0)

## Implementation description
* Followed general format of adminTable

## Steps to test
1. check that the data displayed is correct
2. check that searching/sorting works

## What should reviewers focus on?
* see comments

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: frontend/src/components/pages/AdminPage.tsx [hunk 1]
```diff
@@ -44,7 +45,7 @@ const ErrorState = (): React.ReactElement => (
   </VStack>
 );
 
-const getAdminUser = (user: AdminUser) => {
+const getUser = (user: User) => {
```
Comment (brian386) [inline frontend/src/components/pages/AdminPage.tsx:48]: I changed this function to getUsers instead of creating a new getTeacherUsers function, since I thought they would do the same thing.

File: frontend/src/components/pages/AdminPage.tsx [hunk 2]
```diff
@@ -70,27 +71,40 @@ const AdminPage = (): React.ReactElement => {
...
-  const filteredAdmins = React.useMemo(() => {
-    let filteredUsers = data?.usersByRole;
+  const {
+    loading: teacherLoading,
+    error: teacherError,
+    data: teacherData,
+  } = useQuery(GET_USERS_BY_ROLE, {
+    fetchPolicy: "cache-and-network",
+    variables: { role: "Teacher" },
+  });
```
Comment (brian386) [inline frontend/src/components/pages/AdminPage.tsx]: Adding this slowed the loading time. Would it be better to have a single query getting both teacher and admin roles, and then filtering in the frontend based on which table they're on? Idk if it's worth it though because we might need to change how the backend works.

File: frontend/src/components/user-management/TeacherUserTable.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import React from "react";
```
Comment (brian386) [inline frontend/src/components/user-management/TeacherUserTable.tsx:1]: The ticket said to create a new TeacherTable but it turned out to be identical to adminTable. Would it be better to reuse AdminTable instead?

File: frontend/src/components/user-management/TeacherUserTable.tsx [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import RemoveUserPopover from "./RemoveUserPopover";
+
+interface TeacherUserTableProps {
+  teacherUsers: User[];
+}
+
+const TeacherUserTable = ({
+  teacherUsers,
+}: TeacherUserTableProps): React.ReactElement => {
+  const headers = ["Name", "Email"];
```
Comment (joyce-shi) [inline frontend/src/components/user-management/TeacherUserTable.tsx]: It shouldn't be identical to `AdminTable`! it should also have a `School` field. Please see design for more details!

File: frontend/src/components/pages/AdminPage.tsx (hunk 2 above)
Comment (carissa-tang) [reply frontend/src/components/pages/AdminPage.tsx]: we have a resolver called `users` that you can use to get all users - you will need to define a new query in `APIClients/queries/UserQueries.ts` that returns the needed fields + role

File: frontend/src/components/pages/AdminPage.tsx [hunk 5]
```diff
@@ -164,7 +199,30 @@ const AdminPage = (): React.ReactElement => {
...
+                        />
+                        <InputRightElement pointerEvents="none" h="full">
+                          <SearchOutlineIcon />
+                        </InputRightElement>
+                      </InputGroup>
+                      <SortTablePopover OrderingSets={OrderingSets} />
+                    </HStack>
+                    {search && (
+                      <Text fontSize="16px" color="grey.300" width="100%">
+                        Showing {admins.length} results for &quot;{search}&quot;
```
Comment (carissa-tang) [inline frontend/src/components/pages/AdminPage.tsx]:
```suggestion
                        Showing {teachers.length} results for &quot;{search}&quot;
```
double check that this returns the right values

**Review by carissa-tang — COMMENTED** [review]
are the changes you made to `yarn.lock` necessary? the old dependencies should work with your code changes since i don't think you imported anything new

File: frontend/src/components/user-management/TeacherUserTable.tsx (hunk 4 above)
Comment (carissa-tang) [reply frontend/src/components/user-management/TeacherUserTable.tsx]: looks like there's discrepency in the designs - @brian386 can you follow up with Shawn to see whether it should be a `Region` or `School` field?

image


image

File: frontend/src/components/pages/AdminPage.tsx [hunk 6]
```diff
@@ -164,7 +199,30 @@ const AdminPage = (): React.ReactElement => {
                   </VStack>
                 </TabPanel>
                 <TabPanel>
-                  <AdminUserTable adminUsers={admins} />
+                  <VStack pt={4} spacing={6}>
```
Comment (carissa-tang) [inline frontend/src/components/pages/AdminPage.tsx]: can you refactor the tab panel section to be its own component that accepts either `AdminUserTable` or `TeacherUserTable` as a userTable prop and `admins.length` or `teachers.length` as the count

you may also need to update `SortTablePopover` for teachers (only) to include sorting by `Region`

File: frontend/src/components/user-management/TeacherUserTable.tsx (hunk 4 above)
Comment (brian386) [reply frontend/src/components/user-management/TeacherUserTable.tsx]: Oops sorry I missed that

Comment (brian386) [reply frontend/src/components/user-management/TeacherUserTable.tsx]: Wait is there a good way of getting school data for the user? I don't think the user model has a school field
edit: I think I'm gonna query schools and transform the data manually

Comment (carissa-tang) [reply frontend/src/components/user-management/TeacherUserTable.tsx]: you may want to define a new function in the backend (i.e. `getSchoolByTeacherId`) - might be inefficient to query for each teacher tho

Comment (carissa-tang) [reply frontend/src/components/user-management/TeacherUserTable.tsx]: You can also define a new function that returns all the teacher data with the respective school
-> prob better to keep it in the backend

File: backend/.eslintrc.js [hunk 7]
```diff
@@ -18,5 +18,5 @@ module.exports = {
   rules: {
     "prettier/prettier": ["error", { endOfLine: "auto" }],
   },
-  ignorePatterns: ["build/*"],
+  ignorePatterns: ["build/"],
```
Comment (carissa-tang) [inline backend/.eslintrc.js:21]: curious to know why this change is needed?

File: backend/services/implementations/__tests__/userService.test.ts [hunk 8]
```diff
@@ -90,4 +90,37 @@ describe("mongo userService", (): void => {
 
     expect(emptyRes.length).toEqual(0);
   });
+
+  it("getAllTeachers", async () => {
+    const createRes = await UserModel.insertMany(testUsers);
+
+    const testSchools = [
```
Comment (carissa-tang) [inline backend/services/implementations/__tests__/userService.test.ts:97]: can you use the testSchools in `testUtils/school.ts` directly? then you don't need to define it again and use `testUsers` in `testUtils/users.ts`

File: backend/services/implementations/__tests__/userService.test.ts [hunk 9]
```diff
@@ -90,4 +90,37 @@ describe("mongo userService", (): void => {
...
+        city: "some-city",
+        address: "some-address",
+        teachers: [createRes[1].id],
+      },
+    ];
+
+    await SchoolModel.insertMany(testSchools);
+    const res = await userService.getAllTeachers();
+
+    expect(res[0].school).toEqual("school2");
```
Comment (carissa-tang) [inline backend/services/implementations/__tests__/userService.test.ts]: can you follow the format of the other tests for consistency? i.e. use forEach loop and don't hardcode the expected values

File: backend/services/interfaces/userService.ts [hunk 10]
```diff
@@ -93,6 +99,8 @@ interface IUserService {
    * @param ids the unique identifiers of the users to fetch
    */
   findAllUsersByIds(ids: string[]): Promise<Array<UserDTO>>;
+
+  getAllTeachers(): Promise<TeacherDTO[]>;
```
Comment (carissa-tang) [inline backend/services/interfaces/userService.ts]: add some documentation pls !!

File: backend/.eslintrc.js (hunk 7 above)
Comment (brian386) [reply backend/.eslintrc.js:21]: The linting script was trying to lint my build files and was having a bunch of errors and I had to make that change to fix it. I could remove this change if I'm the only one affected.

File: backend/services/implementations/__tests__/userService.test.ts (hunk 8 above)
Comment (brian386) [reply backend/services/implementations/__tests__/userService.test.ts:97]: I'm not sure if I can use the existing testUsers, because the testUsers don't have an authId, which I need in order to create users in the model. I think these users are intended as return values for mock user service but in this case I need to create the actual users in the database since I'm aggregating.

Comment (carissa-tang) [reply backend/services/implementations/__tests__/userService.test.ts:97]: ohhh ok sounds good then

File: backend/.eslintrc.js (hunk 7 above)
Comment (carissa-tang) [reply backend/.eslintrc.js:21]: hmm that's interesting, i would assume `"build/*"` ignores all build files. could you try deleting your `backend/build` folder and revert the change?

Comment (carissa-tang) [reply backend/.eslintrc.js:21]: if not, i think `"build/"` works for me so it's fine to change it

Comment (brian386) [reply backend/.eslintrc.js:21]: kk I'll just revert it for now

File: backend/graphql/types/userType.ts [hunk 11]
```diff
@@ -41,6 +51,7 @@ const userType = gql`
     users: [UserDTO!]!
     usersByRole(role: String!): [UserDTO!]!
     usersCSV: String!
+    teachers: [TeacherDTO]
```
Comment (carissa-tang) [inline backend/graphql/types/userType.ts:54]:
```suggestion
    teachers: [TeacherDTO!]!
```
assuming we always have teacher users

File: backend/graphql/types/userType.ts [hunk 12]
```diff
@@ -16,6 +16,16 @@ const userType = gql`
     currentlyTeachingJM: Boolean
   }
 
+  type TeacherDTO {
+    id: ID!
+    firstName: String!
+    lastName: String!
+    email: String!
+    grades: [String]
```
Comment (carissa-tang) [inline backend/graphql/types/userType.ts:24]:
```suggestion
    grades: [String!]!
    currentlyTeachingJM: Boolean!
```
i think we can add this as required, given it is a teacher

File: frontend/src/types/UserTypes.ts [hunk 13]
```diff
@@ -1,5 +1,6 @@
-export type AdminUser = {
+export type User = {
```
Comment (carissa-tang) [inline frontend/src/types/UserTypes.ts]: do you mind creating a `AdminUser` and `TeacherUser`? so we can enforce that teacher users have a required school field

File: frontend/src/components/user-management/TeacherUserTable.tsx [hunk 14]
```diff
@@ -0,0 +1,21 @@
+import React from "react";
+import { User } from "../../types/UserTypes";
+import { TableRow, Table } from "../common/Table";
+import RemoveUserPopover from "./RemoveUserPopover";
+import { UserTableProps } from "./AdminTab";
+
+const TeacherUserTable = ({ users }: UserTableProps): React.ReactElement => {
```
Comment (carissa-tang) [inline frontend/src/components/user-management/TeacherUserTable.tsx]:
```suggestion
const TeacherUserTable = ({ users }: TeacherUserTableProps): React.ReactElement => {
```
and add 
interface TeacherUserTableProps {
  users: TeacherUser[];
}

File: frontend/src/components/user-management/AdminUserTable.tsx [hunk 15]
```diff
@@ -1,17 +1,12 @@
 import React from "react";
-import { AdminUser } from "../../types/UserTypes";
+import { User } from "../../types/UserTypes";
 import { TableRow, Table } from "../common/Table";
 import RemoveUserPopover from "./RemoveUserPopover";
+import { UserTableProps } from "./AdminTab";
 
-interface AdminUserTableProps {
```
Comment (carissa-tang) [inline frontend/src/components/user-management/AdminUserTable.tsx:6]: can restore this

File: backend/services/interfaces/userService.ts [hunk 16]
```diff
@@ -93,6 +99,13 @@ interface IUserService {
    * @param ids the unique identifiers of the users to fetch
    */
   findAllUsersByIds(ids: string[]): Promise<Array<UserDTO>>;
+
+  /**
+   * This method gets all teachers with a 'school' field
```
Comment (carissa-tang) [inline backend/services/interfaces/userService.ts:104]:
```suggestion
   * This method gets all teachers and their associated school name
```

File: frontend/src/components/common/SortTablePopover.tsx [hunk 17]
```diff
@@ -0,0 +1,134 @@
+import React from "react";
```
Comment (carissa-tang) [inline frontend/src/components/common/SortTablePopover.tsx:1]: delete this file pls - it was removed in this pr https://github.com/uwblueprint/jump-math/pull/116

File: frontend/src/components/user-management/AdminTab.tsx [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+  InputGroup,
+  InputRightElement,
+  HStack,
+} from "@chakra-ui/react";
+
+import { User } from "../../types/UserTypes";
+
+import { SearchOutlineIcon } from "../../assets/icons";
+
+type UserProperty = "firstName" | "email" | "school";
```
Comment (carissa-tang) [inline frontend/src/components/user-management/AdminTab.tsx]:
```suggestion
```
pay attention when you're fixing merge conflicts! check this file and remove all code and imports that aren't used

File: frontend/src/components/pages/admin/UsersPage.tsx [hunk 19]
```diff
@@ -34,59 +38,108 @@ const ErrorState = (): React.ReactElement => (
...
+const getUser = (user: User) => {
   return {
     email: user.email,
     firstName: user.firstName,
     lastName: user.lastName,
+    school: user.school ? user.school : null,
   };
 };
 
+type UserProperty = "firstName" | "email" | "school";
```
Comment (carissa-tang) [inline frontend/src/components/pages/admin/UsersPage.tsx]:
```suggestion
```
same comment - remove all unused code

**Review by carissa-tang — CHANGES_REQUESTED** [review]

**Review by carissa-tang — APPROVED** [review]
